### ISSUE TITLE
Profiler Agent not picking up Project ID and Zone automatically

### ISSUE THREAD

**Issue by poulad** [GoogleCloudPlatform/cloud-profiler-java#40]
Hi. I am trying to instrument Cloud Profiler for a Spring Boot web app. The agent fails to query Project ID and Zone when it's running in a container on GKE. Everything works fine if I hard code the project ID and zone name in the environment variable. Agent picks up service name and its version automatically correctly.

My Dockerfile is very similar to this setup:
```dockerfile
FROM jetty:9.4.42-jdk11

ADD app.war $JETTY_BASE/webapps/ROOT.war

USER root

ENV JAVA_TOOL_OPTIONS="${JAVA_TOOL_OPTIONS} -agentpath:/opt/cprof/profiler_java_agent.so=-logtostderr"

# -cprof_project_id=my-project,-cprof_service=my-service,-cprof_service_version=0.0.1,-cprof_zone_name=us-central1-c

RUN set -x \
  && mkdir -p /opt/cprof \
  && wget -q -O- "https://storage.googleapis.com/cloud-profiler/java/latest/profiler_java_agent.tar.gz" \
    | tar xzv -C /opt/cprof

USER jetty
```

parts of error logs at container startup:
(If I hard-codes the Project ID, I will get similar failure logs for querying Zone name) 
```
Project ID is not set via flag or environment, will get from the metadata server
Will use profiler service cloudprofiler.googleapis.com to create and upload profiles
Failed to make HTTP request: Couldn't connect to server
Error making HTTP request for 169.254.169.254:80/computeMetadata/v1/project/project-id to the GCE metadata server. Will retry in 1s
Project ID is unknown
Failed to initialize deployment, stop profiling
Exiting the profiling loop
```

I'm not sure if this is an issue with service account's permissions. I can execute the same HTTP request inside the app's GKE container and see the project ID.
```shell
jetty@my-service-deployment-577d7b5fs22v7:~$ curl -fS -H 'Metadata-Flavor: Google' 'http://169.254.169.254:80/computeMetadata/v1/project/project-id'

my-project
```

**Comment by aalexand** [comment]
Seems like either
* connectivity error (the GCE metadata server is not reachable from the container)
  * but you are saying you tried the HTTP request manually from an equivalent environment and this works.
* OR, a race condition where for some reason the GCE metadata server is not reachable seconds after the container start.
  * we saw something like this in the past in the Istio environment where a proxy would be involved.
    * but this can be a red herring

**Comment by poulad** [comment]
Looks like this issue is a duplicate of #4

I am trying whitelisting egress to GCP Metadata server in Istio configs today

https://github.com/GoogleCloudPlatform/microservices-demo/blob/447474f9114b836f121e0dbd65212408c2241aee/istio-manifests/allow-egress-googleapis.yaml#L31-L46

**Comment by poulad** [comment]
@aalexand thank you for the help. It was due to Istio proxy's side-car. It took Istio 7+ seconds to become ready and the profiler was failing before that. I changed the delay between profiler's retries  `-cprof_gce_metadata_server_retry_sleep_sec=9`.

We do not have any blacklist egress rules on Istio in the environment that I'm trying so that `allow-egress-google-metadata` rule had no effects in my case.
